### PR TITLE
Refactoring of template resolving logic

### PR DESCRIPTION
**Pull request by sebkehr**
@demiankatz, @EreMaijala:

This (experimental) PR refers to our recent call last month when we talked about modularization and deals with the mentioned issue about template resolving:

Firstly, modules may now provide additional templates via configuration as described [in the official Laminas docs.](https://docs.laminas.dev/laminas-view/quick-start/#configuration) - especially see configuration key `template_path_stack`. As template folders found in themes take precedence modules may now contain default templates customizable by themes.

Secondly, template paths are now resolved in VuFind-specific ways only if Zend's template injection listener maps a controller to some name prefixed by a value contained in a configurable list - see `vufind.template_injection` in `module/VuFind/config/module.config.php`. This way 3rd-party modules may now safely rely on Zend's default template resolving behavior.

What do you think?

TODO
- [x] Merge with latest master
- [x] Migrate to Laminas
- [x] Discuss configuration of template injection
- [x] Ensure that all existing VuFind modules (including auto-generated custom modules) work correctly with the new mechanism
- [x] Run full test suite
- [x] Update changelog when merging

### CODE REVIEW

**Comment by demiankatz** [comment]
Thanks, @kese, this looks like a great start, and I will do some deeper testing soon. First, though, a couple of questions:

1.) Is there a particular reason you based this against the release-5.1 branch instead of master? Would you object if I take the time to bring it up to date? (I'm guessing this change would probably be targeted for release 7.0 due to the potential for backward incompatibility).

2.) VuFind currently has several modules that render templates in the themes (VuFindAdmin, etc.) -- I assume we'll need to add 'template_injection' keys to them.

3.) Somewhat related to item 2 above, I wonder if we should include a template_injection key in the local module configuration, and make sure that install.php pre-fills it appropriately based on the user-specified namespace; I assume by default we should expect users to want to use VuFind's themes for their local custom code.

In any case, thanks again for getting this started -- and for adding functionality while reducing lines of code! Once I've heard back from you, I'll be happy to help do some more work on this and also begin some more thorough testing.

**Comment by kese** [comment]
Great, @demiankatz. To answer your questions:

1) Simply because our (finc community) code base is still based on the @release-5.1@ branch and I only had little time to spend on this PR. But, please, feel free to adjust my changes to the current master branch.

2) Yes, you are right: `VuFindAdmin/`, `VuFindDevTools/` etc. would have to be added to the list as well.

3) I see your point: Modules generated by the installer should safely expect the usual behavior. Thus, such an entry should be automatically added to the list.

But another option that comes to mind is to invert semantics of the list such that it contains prefixes for controllers depending on Zend's default template resolving behaviour. This way no breaking changes would be introduced and the issues related to 2) and 3) would not arise. 

What do you think?

**Comment by demiankatz** [comment]
@kese, I'll try to work on upgrading this to the latest master when time permits (hopefully in the next week or two).

Regarding the idea of inverting the list semantics, that is certainly worth considering, though I would also welcome input from others. The advantage, as you say, is that it prevents breaking changes. I think the biggest disadvantage is that it feels less self-contained. If we keep it as written here, each module that depends on the themes reports that fact. If we invert it, we'll have to use a different module to disable the theme system for unrelated modules, which feels a little less elegant. I'll try to remember to mention this on the next developers call and see if we can get any more opinions/ideas from others.

**Comment by kese** [comment]
Thanks a lot, @demiankatz.

Concerning 2) the wrong inflection seems to be caused by the usage of back instead of forward slashes. The configuration expects the latter as controller names are first mapped with back slashes being converted to forward slashes and only afterwards inflected.

Concerning 3) yes, I guess we could extract those changes related only to overriding the Zend's default template injection listener to clear things up even further. Do you have a separate pull request in mind?

**Comment by demiankatz** [comment]
Thanks, @kese, I'll fix up the tests next week.

And I'm happy to put together a separate PR to separate the template injection listener overriding changes -- another project for the near future.

**Comment by demiankatz** [comment]
@kese, I have just opened #1436 to separate the code simplification in this PR from the functional changes. Do you mind taking a quick look at that? If you see no problems, I think that can be merged right away, and then we can refine this PR down to just the functional differences.

I've also fixed the broken tests here -- thanks for catching my error with the slashes!

**Comment by demiankatz** [comment]
@kese, I've gone ahead and merged #1436, as I needed those simplifications for some other work. This is now a bit easier to read. We should continue discussing the remaining open issues as time permits.

**Comment by demiankatz** [comment]
@kese @EreMaijala Now that VuFind 6.1 is released and 7.0 development is underway, we should revisit this pull request, decide on an approach, and finish up the work here. Have either of you given this any further thought? Do you want me to do a thorough review and share some new thoughts? Would it be worthwhile to devote some time on a future Community Call (or a separate meeting) to discuss further?

**Comment by EreMaijala** [comment]
@demiankatz I must admit that template resolution is something I haven't really dug into, so I can't say I can provide too much valuable input here. This seems like a good idea, though, and I have nothing against moving this forward. Just one thing: could we have a default `'template_injection' => ['<modulename>/']` for each module, or do you want it to be explicitly expressed in the module configuration?

**Comment by demiankatz** [comment]
@EreMaijala, I think it probably makes the most sense to put an explicit template_injection setting into each individual module that depends upon template injection; that way, the modules are more self-contained, and we don't make any assumptions about third-party components (or even VuFind modules that don't rely on templates, like VuFindConsole).

**Comment by demiankatz** [comment]
@kese, I'd also be interested in your thoughts on the recent conversation here. If we're all on the same page about how to proceed, I can go ahead with Laminas migration here as well.

**Comment by demiankatz** [comment]
@sebkehr, it appears that your username has changed, which perhaps interfered with some past notifications. In any case, I'd love to touch base about the status of this work. The VuFind 7.0 release date is now on the calendar, and it would be great to wrap up this work and merge it sometime this month so it can be included. I'm happy to invest some time, but I'd like to be sure you're available for discussion/approval before I take that step. (In the meantime, I have a few other projects I can work on...)

**Comment by demiankatz** [comment]
Due to lack of activity, I've been unable to resolve this in time for the 7.0 release; I've pushed the milestone down to 8.0 for now.

**Comment by lahmann** [comment]
@demiankatz just to give you a short heads up: we will resume work on this pull request in April - sorry for leaving this untouched for such a long time.

**Comment by demiankatz** [comment]
Thanks, @lahmann -- let me know how I can help, and I'll refresh my memory on where we left this so I'll be ready!

**Comment by demiankatz** [comment]
@lahmann, just checking in to see if any progress has been made here, and whether we should continue to target 8.0 or start looking ahead to 9.0. We can discuss in the PMC meeting on Monday if necessary.

**Comment by demiankatz** [comment]
Thanks for the progress, @domerz. Please let me know if you'd like any help resolving the conflicts that are currently showing, or if you're ready to have this reviewed at some point.

**Comment by domerz** [comment]
@demiankatz we're currently trying to build up a real-life scenario to test the feature with. We hope to integrate the templates provided via one of our VuFind modules from a composer package, to see if it actually works for us. This might take another week or two.

**Comment by demiankatz** [comment]
@domerz, that is a great idea... and I suspect that I may have some use cases for this too in the coming months that might be useful to help with testing/development. In the meantime, let me know if there's anything I can do to help. As noted above, I'm certainly happy to help resolve the conflicts here if that would save you any time.

**Comment by demiankatz** [comment]
Thanks for the progress, @domerz, this is looking pretty good. One question: do we need to add more configuration to other existing modules that depend on the theme system (e.g. VuFindAdmin, VuFindDevTools) for them to work correctly now? I haven't tested, but based on a code read I suspect it may be necessary.

**Comment by demiankatz** [comment]
@domerz, just a check to see if you saw my question above. With the VuFind 8.0 release coming out on 9/20, I'm hoping to have all the code locked by the end of next week. Let me know if there's anything I can do to help test or finalize this.

**Comment by domerz** [comment]
@demiankatz Finally, I managed to include one of our VuFind modules provided as a Composer package. The last commit thus is only for demonstration purposes. The details page of a record should now contain a new Tab Worldcat. Even without proper configuration that should be rendered correctly and try to retrieve some information.
The commit can actually be include in the current dev branch to see that it does indeed not work there.
This should be sufficient to show the PR works properly.

Do you see the need for further work on existing parts of VuFind's code?

**Comment by demiankatz** [comment]
Thanks, @domerz, I will try to find time to test this out later today. My main outstanding question is whether other VuFind modules that use theme templates work, or if we need to add additional `template_injection` configuration for things like VuFindAdmin and VuFindDevTools... but if you're not sure, I can test that out when looking at everything else.

**Comment by demiankatz** [comment]
@domerz, I had time to test this, and I can confirm that your demo works as expected. I also found that, as I suspected, the VuFindDevTools and VuFindAdmin modules (as well as any local custom modules) were broken by the change. I've pushed up fixes to the configuration (as well as some comments to make this a little more clear -- perhaps). It is noteworthy that any user with a custom module containing controllers will need to update the configuration to include the `template_injection` key or else all their custom code will break. I don't think there's any way around this, but it's something we'll need to emphasize heavily in the changelog (unless you can think of a more elegant solution).

In any case, thanks for finishing this up! I think it's ready to merge, unless you have further ideas for improvement or @EreMaijala has other thoughts.

File: module/VuFind/config/module.config.php [hunk 1]
```diff
@@ -547,6 +547,10 @@
     // This section contains all VuFind-specific settings (i.e. configurations
     // unrelated to specific Laminas components).
     'vufind' => [
+        // List of prefixes leading to simpler (non-default) inflection.
```
Comment (EreMaijala) [inline module/VuFind/config/module.config.php]: I must admit I don't understand this comment. There's injection and inflection but the meaning of neither is clear when you don't know how Laminas' default template resolution works. If 'template_injection' is practically an array of path prefixes to strip away, I'd rather name it accordingly, like 'stript_template_path_prefixes'.

File: module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php [hunk 2]
```diff
@@ -49,7 +49,7 @@ class ThemeInjectTemplateListenerTest extends \PHPUnit\Framework\TestCase
      */
     public function testNamespaceStripping()
```
Comment (EreMaijala) [inline module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php]: Should this be more like `testPrefixStripping()`?

File: module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php [hunk 3]
```diff
@@ -49,7 +49,7 @@ class ThemeInjectTemplateListenerTest extends \PHPUnit\Framework\TestCase
...
     {
-        $l = new InjectTemplateListener();
+        $l = new InjectTemplateListener(['VuFind/']);
         $this->assertEquals(
```
Comment (EreMaijala) [inline module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php:54]: The tests should also include one that verifies that something in unlisted namespaces is not changed, such as
```
        $this->assertEquals(
            'vufindadmin/admin',
            $l->mapController(\VuFindAdmin\Controller\AdminController::class)
        );

```

**Review by EreMaijala — CHANGES_REQUESTED** [review]
I took some time to dig into this to understand it properly. Please see the comments. In general, this seems to work, but I would really like to have a backward compatible mechanism such as defaulting to module's name if it doesn't explicitly configure prefixes. I'd also name and comment the config key differently so that however looks at this doesn't need to understand the internals.

Remember to drop the WorldCat commit. :)

File: module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php (hunk 2 above)
Comment (demiankatz) [reply module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php]: Good idea -- done!

File: module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php (hunk 3 above)
Comment (demiankatz) [reply module/VuFindTheme/tests/unit-tests/src/VuFindTest/ThemeInjectTemplateListenerTest.php:54]: Also done.

**Comment by demiankatz** [comment]
Thanks for the progress and conversation, @EreMaijala and @domerz. A few comments and thoughts:

1.) I've improved the tests as @EreMaijala suggested.

2.) I fixed a style issue in the recently-added code. Since @domerz already fixed most of them, it makes me wonder if those fixes were done by hand. If you're fixing these issues manually, please keep in mind that you can run `vendor/bin/phing php-cs-fixer` (and, if necessary, also `vendor/bin/phing phpcbf`) to manually correct most style problems; if you're not already doing that, it will save you some time and trouble!

3.) While I agree that there are major backward-compatibility and user-comprehensibility advantages to using VuFind's theme system for all modules by default, I don't see how we can do this and retain compatibility with third-party Laminas modules. If we use VuFind's custom inflection for every loaded module, then I don't see how we can ever use default Laminas inflection, which means that templates will not load correctly for third-party modules. Maybe this doesn't matter if we only care about building VuFind-specific third-party modules... but if we want generic Laminas modules to work right, we have a problem, because not only will template path prefixes get stripped off, but the actual inflection behavior will also change (since VuFind just lowercases everything, while default Laminas converts camelCase to skewer-case). I think the best we can do is add a module.config.php setting that lists modules which should NOT have VuFind's custom inflection applied to them, and then if a user wants to plug in a standard Laminas module, they need to add a configuration in their local custom module excluding it from custom inflection. That might give us the flexibility we need, while avoiding the need for custom configuration in scenarios where we don't need or want it. I'll take a stab at an implementation now.

**Comment by demiankatz** [comment]
Okay, please take a look at a299b1c -- this adds the ability to add and remove prefixes from consideration for custom inflection, but most things should work by default with no special action needed. I renamed settings in an effort to be more clear, and I added an extensive comment to the VuFind module's configuration explaining the settings. It appears that the custom finc module still works after all of these changes.

We'll definitely want to build a wiki page about adding VuFind extension modules and using default-style Laminas modules (which I haven't actually tested, because I'm not sure if there's a useful example to play with).

In any case, I think we're now more flexible than we were before we started, and we've also made this change much less disruptive.

**Comment by domerz** [comment]
Thank you @EreMaijala and @demiankatz  for your input.
2) I did actually use the automatic code-style-fixer phpcbf and double checked with phpcs. No idea why this didn't find all the issues
3) Maybe a configuration of the "inflectable" (is that a word?) modules might be provided with the VuFindThemes module itself. Obviously without registering the Themes module there is no need to configure which modules to include. I consider it an active decision to list all desired modules in application.config.php. The same decision should then lead to adding the modules' template path prefixes to an allow list (which then reflect a "List of VuFind modules").

**Comment by domerz** [comment]
Do you think we can provide a standard list of modules likely to be excluded?

**Comment by demiankatz** [comment]
@domerz, a couple of comments:

2.) I think what happened is that phpcbf caught most of the issues, but there was one that required php-cs-fixer. There are actually three tools in play here: phpcs detects one set of issues; phpcbf auto-fixes a subset of the issues found by phpcs; php-cs-fixer auto-fixes a different set of issues from phpcs + phpcbf. Sorry that it's all a bit confusing!

3.) Unfortunately, I don't think we can view application.config.php as the authoritative place where modules and prefixes are listed, because that is just the baseline, and then the VUFIND_LOCAL_MODULES environment variable adds more. This is necessary to support multi-site instances, where different sets of modules may be loaded conditionally based on different Apache configurations. I suppose we could add one or more new environment variables to also control template behavior, but that feels like overkill to me at this stage. For me, the bottom line is that most people aren't going to know or care about (or understand) this mechanism, and I would prefer to keep it fairly invisible except in those cases where it is actually needed.

I'm certainly open to moving the example configuration and documentation from the VuFind module's module.config.php somewhere else -- that could certainly go in application.config.php to show where configuration belongs. Perhaps it even makes sense to start with a default `excluded_theme_prefixes` value of `'Laminas'` (without a trailing slash), since that will protect us against inappropriately inflecting any core Laminas modules that might happen to provide templates. (I don't think we need it right now, but it would be a harmless example).

I do agree that it could also make sense to put this in the VuFindTheme module, but right now, there's not a logical place to put it without restructuring the way that module exposes configuration (since it uses methods of the module class instead of a module.config.php file for its configs).

**Comment by demiankatz** [comment]
Well, never mind on the idea of moving configs to application.config.php -- that doesn't seem to work. Must have to do with the order of operations.

**Comment by demiankatz** [comment]
@domerz, I put a bit more work into this with commit 8365209 -- I moved the configuration into the VuFindTheme module as you suggested. I also made the exclude functionality a little smarter (so that it's now possible to exclude all the Laminas modules with a setting of "Laminas"). What do you think?

**Comment by domerz** [comment]
@demiankatz I like that solution very much and see no reason how this should break with existing modules.

**Comment by demiankatz** [comment]
Thanks, @domerz! I noticed that the factory for the InjectTemplateListener was built in a non-standard way, and since it contains some complex logic, it needed test coverage... so I've modernized it and added a unit test. I'm now running the full test suite to be sure there are no unanticipated side effects. If that passes, I believe this is ready to merge, but I'll wait to hear if @EreMaijala has any last-minute suggestions first. I expect to be able to merge this tomorrow, if not sooner.

**Review by EreMaijala — APPROVED** [review]
Thanks for all the work on this. I think this looks very good now!

**Comment by demiankatz** [comment]
All tests are passing -- merging now!

**Pull request merged**